benchmarks/benchmarks/sparse_linalg_svds.py

```diff
@@ -24,8 +31,17 @@ def setup(self, k, problem, solver):
                                 "svds_benchmark_files.npz")
         matrices = np.load(datafile, allow_pickle=True)
         self.A = matrices[problem][()]
+        _, s, _ = svd(self.A.toarray(), full_matrices=False)
+        self.all_singular_values = s.sort()
 
     def time_svds(self, k, problem, solver):
-        # consider k = int(np.min(self.A.shape) * k)
-        np.random.seed(0)
-        svds(self.A, k=k, solver=solver)
+        rng = np.random.default_rng(0)
+        if solver == 'svd':
+            _, s, _ = svd(self.A.toarray(), full_matrices=False)
+        else:
+            _, s, _ = svds(self.A, k=k, solver=solver, random_state=rng)
+
+        tol = k * np.prod(self.A.shape) * np.finfo(float).eps
+        k_singular_values = self.all_singular_values[:k]
+        accuracy = max(abs(k_singular_values - s) / k_singular_values)
+        assert accuracy < tol, msg
```

Comment:
I'm probably resurrecting a really long discussion, and I think there was a mailing list post as well, but can you briefly summarize the policy decision on putting `assert` directly in benchmarks like this? I'm sure a bunch of people must have said stuff like "that belongs in unit tests," etc.

I'm a bit more curious about the design decision though--for example, if an algorithm is not deterministically decisive in finding a solution, perhaps it might benefit from a kwarg that allows the function to raise an error when it fails to converge within some tol? Perhaps I'm just digging for a reason to avoid putting asserts in the bench suite though...

Reply:
If the point of the `assert` is to keep the benchmark in a working state, then it's fine. It's not a replacement for a unit test to ensure that the function being benchmarked is working.

For example, we may intentionally change some defaults that will change what this particular invocation of `svds()` does in fine detail, maybe it will cause it to use fewer iterations and exit before reaching a useful level of accuracy. Then we want to come in and change the invocation of `svds()` to restore that level of accuracy so that we're benchmarking the same thing (i.e. "run `svds()` on this problem until we get a useful level of accuracy") over time.

If instead, we come in and decide to change `svds()` itself, then we're abusing the benchmark as a unit test, and need to expand our unit test suite.

Reply:
I see, I suppose it may be slightly annoying to maintain a test like `test_svds_accuracy_with_defaults`, though that might also be viable. I'm not sure if having to write unit tests that mirror the benchmarks is much better than a plain `assert` here though, since the maintenance connection could be more annoying.